<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>RunnableFutureCallbackAdapterTest.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">threadly$All_in_threadly_test.exec</a> &gt; <a href="index.source.html" class="el_package">org.threadly.concurrent.future</a> &gt; <span class="el_source">RunnableFutureCallbackAdapterTest.java</span></div><h1>RunnableFutureCallbackAdapterTest.java</h1><pre class="source lang-java linenums">package org.threadly.concurrent.future;

import static org.junit.Assert.*;

import java.util.concurrent.CancellationException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeoutException;

import org.junit.Test;
import org.threadly.ThreadlyTester;
import org.threadly.test.concurrent.AsyncVerifier;
import org.threadly.test.concurrent.TestCondition;

@SuppressWarnings({&quot;javadoc&quot;, &quot;deprecation&quot;})
<span class="fc" id="L15">public class RunnableFutureCallbackAdapterTest extends ThreadlyTester {</span>
  @Test
  public void constructorTest() {
<span class="fc" id="L18">    Future&lt;Object&gt; future = new TestFutureImp(false);</span>
<span class="fc" id="L19">    FutureCallback&lt;Object&gt; callback = new TestFutureCallback();</span>
<span class="fc" id="L20">    RunnableFutureCallbackAdapter&lt;?&gt; rfca = new RunnableFutureCallbackAdapter&lt;&gt;(future, callback);</span>
    
<span class="pc bpc" id="L22" title="1 of 2 branches missed.">    assertTrue(rfca.future == future);</span>
<span class="pc bpc" id="L23" title="1 of 2 branches missed.">    assertTrue(rfca.callback == callback);</span>
<span class="fc" id="L24">  }</span>
  
  @Test
  @SuppressWarnings(&quot;unused&quot;)
  public void constructorFail() {
    try {
<span class="nc" id="L30">      new RunnableFutureCallbackAdapter&lt;&gt;(null, new TestFutureCallback());</span>
<span class="nc" id="L31">      fail(&quot;Exception should have thrown&quot;);</span>
<span class="fc" id="L32">    } catch (IllegalArgumentException e) {</span>
      // expected
<span class="nc" id="L34">    }</span>
    try {
<span class="nc" id="L36">      new RunnableFutureCallbackAdapter&lt;&gt;(new TestFutureImp(false), null);</span>
<span class="nc" id="L37">      fail(&quot;Exception should have thrown&quot;);</span>
<span class="fc" id="L38">    } catch (IllegalArgumentException e) {</span>
      // expected
<span class="nc" id="L40">    }</span>
<span class="fc" id="L41">  }</span>
  
  @Test
  public void resultTest() {
<span class="fc" id="L45">    SettableListenableFuture&lt;Object&gt; future = new SettableListenableFuture&lt;&gt;();</span>
<span class="fc" id="L46">    TestFutureCallback callback = new TestFutureCallback();</span>
<span class="fc" id="L47">    Object result = new Object();</span>
<span class="fc" id="L48">    future.setResult(result);</span>
    
<span class="fc" id="L50">    RunnableFutureCallbackAdapter&lt;?&gt; rfca = new RunnableFutureCallbackAdapter&lt;&gt;(future, callback);</span>
<span class="fc" id="L51">    rfca.run();</span>
    
<span class="pc bpc" id="L53" title="1 of 2 branches missed.">    assertTrue(callback.getLastResult() == result);</span>
<span class="fc" id="L54">    assertNull(callback.getLastFailure());</span>
<span class="fc" id="L55">  }</span>
  
  @Test
  public void executionExceptionTest() {
<span class="fc" id="L59">    SettableListenableFuture&lt;Object&gt; future = new SettableListenableFuture&lt;&gt;();</span>
<span class="fc" id="L60">    TestFutureCallback callback = new TestFutureCallback();</span>
<span class="fc" id="L61">    Exception failure = new Exception();</span>
<span class="fc" id="L62">    future.setFailure(failure);</span>
    
<span class="fc" id="L64">    RunnableFutureCallbackAdapter&lt;?&gt; rfca = new RunnableFutureCallbackAdapter&lt;&gt;(future, callback);</span>
<span class="fc" id="L65">    rfca.run();</span>
    
<span class="pc bpc" id="L67" title="1 of 2 branches missed.">    assertTrue(callback.getLastFailure() == failure);</span>
<span class="fc" id="L68">    assertNull(callback.getLastResult());</span>
<span class="fc" id="L69">  }</span>
  
  @Test
  public void cancellationExceptionTest() {
<span class="fc" id="L73">    Future&lt;Object&gt; future = new TestFutureImp(false);</span>
<span class="fc" id="L74">    TestFutureCallback callback = new TestFutureCallback();</span>
<span class="fc" id="L75">    future.cancel(true);</span>
    
<span class="fc" id="L77">    RunnableFutureCallbackAdapter&lt;?&gt; rfca = new RunnableFutureCallbackAdapter&lt;&gt;(future, callback);</span>
<span class="fc" id="L78">    rfca.run();</span>
    
<span class="fc" id="L80">    assertTrue(callback.getLastFailure() instanceof CancellationException);</span>
<span class="fc" id="L81">    assertNull(callback.getLastResult());</span>
<span class="fc" id="L82">  }</span>
  
  @Test
  public void interruptedExceptionTest() throws InterruptedException, TimeoutException {
<span class="fc" id="L86">    SettableListenableFuture&lt;Object&gt; future = new SettableListenableFuture&lt;&gt;();</span>
<span class="fc" id="L87">    TestFutureCallback callback = new TestFutureCallback();</span>
<span class="fc" id="L88">    final AsyncVerifier av = new AsyncVerifier();</span>
    
<span class="fc" id="L90">    final RunnableFutureCallbackAdapter&lt;?&gt; rfca = new RunnableFutureCallbackAdapter&lt;&gt;(future, callback);</span>
<span class="fc" id="L91">    final Thread t = new Thread(new Runnable() {</span>
      @Override
      public void run() {
        try {
<span class="nc" id="L95">          rfca.run();</span>
<span class="nc" id="L96">          av.fail(&quot;Exception did not throw&quot;);</span>
<span class="fc" id="L97">        } catch (RuntimeException expected) {</span>
<span class="fc" id="L98">          av.signalComplete();</span>
<span class="nc" id="L99">        }</span>
<span class="fc" id="L100">      }</span>
    });
    
<span class="fc" id="L103">    t.start();</span>
<span class="fc" id="L104">    new TestCondition(() -&gt; t.isAlive()).blockTillTrue();</span>
    
<span class="fc" id="L106">    t.interrupt();</span>
    
<span class="fc" id="L108">    av.waitForTest();</span>
<span class="fc" id="L109">    assertTrue(callback.getLastFailure() instanceof InterruptedException);</span>
<span class="fc" id="L110">    assertNull(callback.getLastResult());</span>
<span class="fc" id="L111">  }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.1.201803210924</span></div></body></html>